relay-metrics/src/aggregator/inner.rs

```diff
@@ -0,0 +1,926 @@
+                });
+            }
+        }
+
+        let slot_diff = SlotDiff {
+            min: config
+                .max_secs_in_past
+                .map_or(u64::MAX, |v| v.div_ceil(u64::from(bucket_interval))),
+            max: config
+                .max_secs_in_future
+                .map_or(u64::MAX, |v| v.div_ceil(u64::from(bucket_interval))),
+        };
+
+        let total_slots = slots.len();
+        Self {
+            slots: VecDeque::from(slots),
+            num_partitions,
+            head: config.start.as_secs() / u64::from(bucket_interval) * u64::from(num_partitions),
+            bucket_interval,
+            delay: config.delay,
+            stats: stats::Total::default(),
+            limits: stats::Limits {
+                max_total: config.max_total_bucket_bytes.unwrap_or(u64::MAX),
+                // Break down the maximum project cost to a maximum cost per partition.
+                max_partition_project: config
+                    .max_project_key_bucket_bytes
+                    .map(|c| c.div_ceil(total_slots as u64))
+                    .unwrap_or(u64::MAX),
+            },
+            slot_diff,
+            partition_by: FlushBatching::Partition,
+            hasher: build_hasher(),
+        }
+    }
+
+    /// Returns the configured bucket interval.
+    pub fn bucket_interval(&self) -> u32 {
+        self.bucket_interval
+    }
+
+    /// Returns total aggregator stats.
+    pub fn stats(&self) -> Stats {
+        Stats {
+            count: self.stats.count,
+            count_by_namespace: self.stats.count_by_namespace,
+            cost: self.stats.count,
+            cost_by_namespace: self.stats.cost_by_namespace,
+        }
+    }
+
+    /// Returns `true` if the aggregator contains any metric buckets.
+    pub fn is_empty(&self) -> bool {
+        self.stats.cost == 0
+    }
+
+    /// Returns the time as a duration since epoch when the next flush is supposed to happen.
+    pub fn next_flush(&self) -> Duration {
+        // `head + 1` to get the end time of the slot not the start.
+        (Duration::from_secs(self.head + 1) / self.num_partitions * self.bucket_interval)
+            + Duration::from_secs(u64::from(self.delay))
+    }
+
+    /// Merges a metric bucket.
+    ///
+    /// Returns `true` if the bucket already existed in the aggregator
+    /// and `false` if the bucket did not exist yet.
+    pub fn merge(
+        &mut self,
+        mut key: BucketKey,
+        value: BucketData,
+    ) -> Result<(), AggregateMetricsError> {
+        let project_key = key.project_key;
+        let namespace = key.metric_name.namespace();
+
+        let time_slot = key.timestamp.as_secs() / u64::from(self.bucket_interval);
+        // Make sure the timestamp is normalized to the correct interval as well.
+        key.timestamp = UnixTimestamp::from_secs(time_slot * u64::from(self.bucket_interval));
+
+        let now_slot = self.head / u64::from(self.num_partitions);
```

Comment:
Should `self.num_partitions` be a `u64` to begin with?

Reply:
It's not, mainly for the division needed in `next_flush`, we can freely extend a `u32` to `u64`, but the other way around requires a cast/is fallible.